README.md
# LiquidCrystal NKC for Arduino IDE

[![GitHub release (latest by date including pre-releases)](https://img.shields.io/github/v/release/domiluci/LiquidCrystal_NKC?label=Release&display_name=tag&include_prereleases)](https://github.com/domiluci/LiquidCrystal_NKC/releases)
[![GitHub](https://img.shields.io/github/license/domiluci/LiquidCrystal_NKC?label=License)](https://github.com/domiluci/LiquidCrystal_NKC/blob/main/LICENSE)
[![Arduino 2.0](https://img.shields.io/badge/Arduino%202.0-Ready-brightgreen)](https://www.arduino.cc/en/software#experimental-software)
[![Donate](https://img.shields.io/badge/Donate-BTC-f7931a?logo=bitcoin)](https://i.ibb.co/QQ4hSnN/CA-BTC-DL.png)
[![Donate](https://img.shields.io/badge/Donate-BCH-0ac18e?logo=bitcoincash)](https://i.ibb.co/DW9gXPN/BP-BCH-DL.png)
[![Donate](https://img.shields.io/badge/Donate-LTC-a6a9aa?logo=litecoin)](https://i.ibb.co/G2y9GXh/BP-LTC-DL.png)
[![Donate](https://img.shields.io/badge/Donate-ETH-434976?logo=ethereum)](https://i.ibb.co/wrbMGZN/BP-ETH-DL.png)

![Hello World Example](https://i.ibb.co/HYcRqPY/Hello-World-2.jpg)

## DESCRIPTION

While not very mainstream, the **[Longtech & NKC Electronics Serial LCD display module](http://media.nkcelectronics.com/datasheet/LCM2004SD-NSA-FBW.pdf)** boasts a myriad of features over and above the generic **PCF8574** and **MCP23017** I2C backpacks typically used to simplify 4-bit & 8-bit LCD connections, such as the capability to command the display not just via I2C but also over RS232 (with just a single wire), as well as over an everyday SPI bus. This NKC LCD also provides completely digital control over the brightness and contrast, doing away with trim pots for contrast adjustment, and providing backlight settings beyond just "ON" and "OFF".

But the bragging rights go beyond the mere NKC display itself. This **LiquidCrystal NKC** library, designed by professional engineering scientist and developer of the groundbreaking Neo Leaf load control system, **[Dominic Luciano](https://www.github.com/domiluci)**, is built to work with the same commands as the **LiquidCrystal** and **LiquidCrystal I2C** libraries, but it also retains alias commands to remain compatible with the original **SerialLCD** library. So little to no code will require rewriting or reworking.

However, many extensions to the basic commands were integrated into this **LiquidCrystal NKC** library. And there are many future additions and expansions planned, so make sure to keep your **LiquidCrystal NKC** library up-to-date to take advantage of all those features!

**NOTE:** This is currently a *"Version 0"* beta release. As such, some features are still being developed and perfected, and therefore may not work properly or as expected. Please make sure to post and report all bugs on the [official GitHub repository](https://www.github.com/domiluci/LiquidCrystal_NKC). Thanks!

### LICENSE

The **LiquidCrystal NKC** library, copyright (C) 2021-2022 Dominic M. Luciano & Nox Labs, is licensed under the **[GNU Lesser General Public License v3.0 (GNU LGPLv3)](https://www.gnu.org/licenses/lgpl-3.0)**, and as such is free for all to use for any purpose, inluding commercial and for profit use. However, it is humbly asked that the original source code, copyright information, and license be disclosed and included in all derivative works, and that all changes made over and above the original source code be documented and disclosed in those derivative works. In short, it would just mean a lot if credit was given when using the **LiquidCrystal NKC** library for Arduino IDE as word-of-mouth is the only way people will ever know this library exists for their use. Thank you all so very much!

## INSTALLATION

### ARDUINO

* **Arduino Library Manager**: 
	* Open your copy of the [Arduino IDE](https://www.arduino.cc/en/software). Go to **Sketch > Include Library > Manage Libraries**
	* Search for **LiquidCrystal NKC** and click **Install**
	* It's that easy! See the following information for instructions on how to include and use **LiquidCrystal NKC**.
* **Arduino ZIP Installation**:
	* Download the **[Latest ZIP Release](https://github.com/domiluci/LiquidCrystal_NKC/releases)** of the **LiquidCrystal NKC** library.
	* Open your copy of the [Arduino IDE](https://www.arduino.cc/en/software). On the toolbar, click **Sketch > Include Library > Add .ZIP Library**
	* In the explorer window, navigate to the **LiquidCrystal NKC** ZIP file you just downloaded and select it.
	* You're all set! See the following information for instructions on how to include and use **LiquidCrystal NKC**.
	
### MANUAL

* **Manual Installation**:
	* Download the **[Latest ZIP Release](https://github.com/domiluci/LiquidCrystal_NKC/releases)** of the **LiquidCrystal NKC** library.
	* Extract the files of that ZIP archive somewhere that's easily accessible. You'll be moving them in a moment.
	* Open your Arduino **"libraries"** folder in **Windows Explorer**, OSX's **Finder**, or other file explorer
	* Once there, make a new folder in **"libraries"** specifically called **"LiquidCrystal_NKC"** and open it up.
	* Copy all of the files you extracted from the ZIP archive into that **"LiquidCrystal_NKC"** folder you just made.
	* The **LiquidCrystal NKC** library should automatically register with the Arduino IDE on next start or restart

## GENERAL USE

### CONNECTIONS

* **I2C LCD Connection to Arduino Uno/Nano**
	* LCD Pin 2 [VSS] -> Arduino GND
	* LCD Pin 3 [VDD] -> Arduino +5V
	* LCD Pin 7 [SCL] -> Arduino A5 [SCL]
	* LCD Pin 8 [SDA] -> Arduino A4 [SDA]
	
* **RS232 LCD Connection to Arduino Uno/Nano**
	* LCD Pin 1 [RXD] -> Arduino D1 [TXD]
	* LCD Pin 2 [VSS] -> Arduino GND
	* LCD Pin 3 [VDD] -> Arduino +5V
* **SPI LCD Connection to Arduino Uno/Nano**
	* LCD Pin 2 [VSS] -> Arduino GND
	* LCD Pin 3 [VDD] -> Arduino +5V
	* LCD Pin 5 [SPISS] -> Arduino D10 [SS or CS]
	* LCD Pin 6 [SDO] -> Arduino D11 [MOSI]
	* LCD Pin 7 [SCK] -> Arduino D13 [SCK]
	* LCD Pin 8 [SDI] -> Arduino D12 [MISO]
	
### INCLUDE

Including **LiquidCrystal NKC** in an Arduino sketch:
	
* Add `#include <Wire.h>` and `#include <LiquidCrystal_NKC.h>` to the top of your `.ino` or `.pde` file.
* Or go to **Sketch > Include Library > Wire** and **Sketch > Include Library > LiquidCrystal NKC**

### CONSTRUCTOR

**Format:** LC_NKC lcd(Interface, I2C Address/Baud Rate, LCD Columns, LCD Rows)

Ex: `LC_NKC lcd(I2C, 0x28, 20, 4);`

Ex: `LC_NKC lcd(RS232, 9600, 20, 4);`

### COMMANDS
* **LiquidCrystal Compatible Commands:**
	* Send a direct HD47780 command: `.command(value)`
	* Initialize the LCD module: `.begin()` or `.init()`
	* Turn the display ON/OFF: `.display()` or `.noDisplay`
	* Turn the backlight ON/OFF: `.backlight()` or `.noBacklight()`
	* Clear the entire display: `.clear()`
	* Reset the cursor to 0, 0: `.home()`
	* Set the cursor's position: `.setCursor(column, row)`
	* Write to the LCD display: `.write()`
	* Print a char or string:  `.print()`
	* Store a unique character: `.createChar()`
	* Set cursor visibility: `.cursor()` and `.noCursor()`
	* Blinking Cursor ON/OFF: `.blink()` and `.noBlink()`
	* Shift entire display LEFT: `.scrollDisplayLeft()`
	* Shift entire display RIGHT: `.scrollDisplayRight()`
	
* **LCD API 1.0 Compatible Commands**
	* Send a direct HD47780 command: `.command(value)`
	* Turn the pixels/backlight ON/OFF: `.on()` or `.off()`
	* Initialize the LCD module: `.init()`
	* Set backlight brightness: `.setBacklight(0 to 255)`
	* Set the display's contrast: `.setContrast(0 to 255)` *(Default = 200)*
	* Clear the entire display: `.clear()`
	* Write to the LCD display: `.write()`
	* Print a char or string:  `.print()`
	* Deprecated print command: `.printstr()`

* **Alias and Other Commands:**
	* Turn display's pixels ON/OFF: `.displayOn()` and `.displayOff()`
	* Set backlight brightness: `.setBacklightBrightness(0 to 255)`
	* Clear a single character: `.clear(column, row)`
	* Clear several characters: `.clear(start column, end column, row)`
	* Clear an entire line/row: `.clear(row)`
	* Set cursor visibility: `.cursorOn()` and `.cursorOff()`
	* Blinking cursor ON/OFF: `.blinkOn()` and `.blinkOff()`
	* Blinking cursor ON/OFF: `.blink(true)` and `.blink(false)`
	* Move cursor LEFT/RIGHT once: `.left()` or `.right()`
	* Move cursor L/R by "columns": `.left(columns)` or `.right(columns)`
	* Backspace the cursor: `.backspace()`
	* Shift the entire LCD once: `.shiftLeft()` and `.shiftRight()`
	* Shift entire LCD by "columns": `.shiftLeft(columns)` and `.shiftRight(columns)`
	* Shift entire LCD LEFT by "columns": `.scrollDisplayLeft(columns)`
	* Shift entire LCD RIGHT by "columns": `.scrollDisplayRight(columns)`

	
## EXAMPLES

### HELLO WORLD

The **Hello World** example sketch for Arduino provide a foundation of basic commands to build on with the **Full Demo** sketch.

* **[Hello World Arduino Example Sketch: I2C Version](https://github.com/domiluci/LiquidCrystal_NKC/blob/main/examples/HelloWorld_I2C/HelloWorld_I2C.ino)**
* **[Hello World Arduino Example Sketch: RS232 Version](https://github.com/domiluci/LiquidCrystal_NKC/blob/main/examples/HelloWorld_RS232/HelloWorld_RS232.ino)**

### FULL DEMO

The **Full Demo** sketch builds on the basics of the **Hello World** sketch while demonstrating advanced commands and alternate uses.

* **[Full Demo Arduino Example Sketch: I2C Version](https://github.com/domiluci/LiquidCrystal_NKC/blob/main/examples/FullDemo_I2C/FullDemo_I2C.ino)**
* **[Full Demo Arduino Example Sketch: RS232 Version](https://github.com/domiluci/LiquidCrystal_NKC/blob/main/examples/FullDemo_RS232/FullDemo_RS232.ino)**

## WHAT'S NEXT

### SHORT TERM

One of the biggest goals of the **LiquidCrystal NKC** library for the Arduino IDE is to be as compatible as possible with existing libraries, such as with the **LiquidCrystal** and **SerialLCD** parent libraries that **LiquidCrystal NKC** was born from. So to expand on that compatibility even more so that users migrating from other libraries can simply "plug-n-play", **LiquidCrystal NKC** is in the process of being made as [**LCD API 1.0**](https://playground.arduino.cc/Code/LCDAPI/) compliant as possible. This means less hassle when switching from one display and library to the **LiquidCrystal NKC** and **NKC Electronics** serial LCD display module, and less rewriting and redesigning to switch back, not that you'll want to, *of course*! So make sure to keep your copy of **LiquidCrystal NKC** up-to-date to capitalize on all the features being added each Friday these coming weeks, especially while all the kinks are getting ironed out in this "Version 0" release.

### LONG TERM

Currently, an expanded **LiquidCrystal** library is in development that will combine support for all LCD displays, including this NKC Serial LCD display. A universal LCD library will not only make it easier to switch between different LCD displays while eliminating time wasted on reconfiguring and recoding projects, but it will also create a standardized set of commands for all LCD displays of all interface types. Not to mention expand on all current commands, providing easier, simpler functionality with more power and flexibility, all while requiring less code to implement and execute, especially for advanced commands.

As for this **LiquidCrystal NKC** library, it will continue to be updated and maintained, even after completion of the expanded **LiquidCrystal** library. More additions will be coming soon. So make sure you keep your copy of the **LiquidCrystal NKC** library up-to-date to score on all those fresh features. And stay tuned to the other [repositories](https://github.com/domiluci?tab=repositories) for the all new, all universal **LiquidCrystal** library coming soon!

## DONATIONS


CHANGELOG.md
# CHANGE LOG

The official and complete change log of the **LiquidCrystal NKC** library for the Arduino IDE.

## VERSION 0

* **v0.21.316** (November 12th, 2021)
	* Changed value of `I2C_SLOW` from `10000` (10kHz) to `50000` (50kHz) as it broke I2C functionality of the LCD when used
	* `on()` and `off()` now turn the pixels *and* backlight **ON** or **OFF**, and are LCD API 1.0 compliant
	* The `setBacklight()` function now accepts a value from 0-255, and therefore is LCD API 1.0 compliant.
	* The `setContrast()` function now accepts a value from 0-255, and therefore is LCD API 1.0 compliant.
	* The **new** `displayOn()` and `displayOff()` commands turn *just the pixels* **ON** or **OFF**
	* Combined all `delayMicroseconds(I2CDELAY)` calls to just one after `Wire.endTransmission()` in `LiquidCrystal_NKC.cpp`
	* Removed unneeded inclusions of `delayMicroseconds(RS232DELAY)` in `LiquidCrystal_NKC.cpp`as they *may* break functionality
	* Further increased the command responsiveness of I2C and RS232 command functions in `LiquidCrystal_NKC.cpp`
	* Resolved the formatting issues of `LiquidCrystal_NKC.h` and `LiquidCrystal_NKC.cpp`
	* Updated the content of **keywords.txt**
	* Updated the content of **library.json**, and **library.properties**
	* Updated the **Commands** and **What's Next** portion of the **[README.md](https://github.com/domiluci/LiquidCrystal_NKC/blob/main/README.md)** file
	* Temporarily removed the **compiled** examples folders to add an auto-compiler to the repository

* **v0.21.309** (November 5th, 2021)
	* Modified variable types to optimize and reduce memory consumption
	* Fixed the overflow warnings for the I2C clock speed constants
	* Increased I2C command response speed by 200%
	* Revised the format and content of **keywords.txt**
	* Revised the content of **library.json**, and **library.properties**
	* Added **[CHANGELOG.md](https://github.com/domiluci/LiquidCrystal_NKC/blob/main/CHANGELOG.md)** to the repository; the official change log
	* Added **Connections** information for Uno/Nano to the **[README.md](https://github.com/domiluci/LiquidCrystal_NKC/blob/main/README.md)** file
	* Added **[LICENSE](https://github.com/domiluci/LiquidCrystal_NKC/blob/main/LICENSE)** file, licensing **LiquidCrystal NKC** under **[GNU LGPLv3](https://www.gnu.org/licenses/lgpl-3.0)**
	
* **v0.21.302** (October 29th, 2021)
	* Official library release and initial GitHub commit


src/LiquidCrystal_NKC.cpp
////////////////////////////////////////////////////////////////////////////////////////////////////
// LiquidCrystal NKC
// Version 0.21.316 (Nov. 12th, 2021)
// Developed by Dominic Luciano
//
// An Arduino library for the NKC Electronics Serial LCD Display running in RS232, I2C, or SPI mode.
// Adapted from the original "SerialLCD" library by NKC Electronics and reworked for compatibility
// with the "LiquidCrystal" family to allow seamless transition between libraries and LCD displays.
//
// Date Codes Generated By: https://www.epochconverter.com/daynumbers
//
// DEFAULT CONNECTION GUIDE ////////////////////////////////////////////////////////////////////////
//
// RS232 (UART) INTERFACE:
// LCD PIN 1 -> ARDUINO UNO DIGITAL PIN 1 [TXD] (or any SoftSerial library compatible pin)
//
// I2C INTERFACE:
// LCD PIN 7 -> ARDUINO UNO ANALOG PIN 5 [SCL]
// LCD PIN 8 -> ARDUINO UNO ANALOG PIN 4 [SDA]
//
// SPI INTERFACE:
// LCD PIN 5 -> ARDUINO UNO DIGITAL PIN 10 [SS or CS]
// LCD PIN 6 -> ARDUINO UNO DIGITAL PIN 11 [MOSI]
// LCD PIN 7 -> ARDUINO UNO DIGITAL PIN 13 [SCK]
// LCD PIN 8 -> ARDUINO UNO DIGITAL PIN 12 [MISO] (Not Necessary)
//
// USAGE:
// Please see the examples in the "examples" folder of this library distibution, or through the
// library examples section of the Arduino IDE. The README file also contains instructions for
// general usage of this library. Enjoy!
//
// Feel free to browse the GitHub repository for this library, or start a discussion at:
// https://www.github.com/domiluci/LiquidCrystal_NKC
//
// CHANGE LOG //////////////////////////////////////////////////////////////////////////////////////
//
// 21.302: Initial Release and GitHub Commit of Entire LiquidCrystal NKC library v0.21.302
// 21.309: Improved I2C Responsiveness and Variable Memory Use. Fixed Overflow Warnings.
// 21.316: Fixed bug when changing clock speed to 10kHz. I2C_SLOW is now 50000. Removed extra delays
//
////////////////////////////////////////////////////////////////////////////////////////////////////

// Include the Library's ".h" File
#include "LiquidCrystal_NKC.h"

// Default LCD Lines and Columns (Do NOT Change Here!)
uint8_t g_num_lines = 2;  // Default: 2 Lines/Rows
uint8_t g_num_col = 16;   // Default: 16 Columns

// Default LCD Interface, I2C Address, I2C Clock, RS232 Baud Rate, and SPI Transfer Mode
uint8_t g_interface = 0;              // 0 = Serial/RS232; 1 = I2C; 2 = SPI
uint8_t g_i2c_address = 0x28;         // Default for NKC I2C LCD = 0x28
uint32_t g_i2c_clock = I2C_STANDARD;  // I2C_SLOW = 10K; I2C_STANDARD = 100K; I2C_FAST = 400K
uint16_t g_baudrate = 9600;           // Default NKC RS232 Baud Rate is 9600bps

// Default Pins (Arduino Uno/Nano)
uint8_t g_txPin = 1;
uint8_t g_sspin = 10;
uint8_t g_sclk = 13;
uint8_t g_mosi = 11;
uint8_t g_miso = 12;

// Hexadecimal Index for Controlling the LCD's Cursor by Line & Column
// NKC Electronics Specific Index
uint8_t g_index[4] = { 0x00, 0x40, 0x14, 0x54 };

////////////////////////////////////////////////////////////////////////////////////////////////////
// CONSTRUCTOR /////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////

// This is the Default Constructor //
// RS232 Example: "LC_NKC lcd(RS232, 9600, 20, 4);"
// I2C Example: "LC_NKC lcd(I2C, 0x28, 20, 4);"
// SPI Example: This functionality is still under development
LC_NKC::LC_NKC(uint8_t interface, uint16_t parameter, uint8_t num_col, uint8_t num_lines) {

  g_num_lines = num_lines;
  g_num_col = num_col;
  g_interface = interface;

  switch (g_interface) {
    case RS232:
      g_baudrate = parameter;
      break;
    case I2C:
      g_i2c_address = parameter;
      break;
    case SPI:
      g_sspin = parameter;
      break;
  }

  // Default to 2 lines/rows if less than 1 or greater than 4
  if (g_num_lines < 1 || g_num_lines > 4) {
    g_num_lines = 2;
  }

  // Default to 16 columns if less than 1 or greater than 40
  if (g_num_col < 1 || g_num_col > 40) {
    g_num_col = 16;
  }
}

////////////////////////////////////////////////////////////////////////////////////////////////////
// HARDWARE LEVEL FUNCTIONS ////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////

// INTERFACE TRANSFER FUNCTIONS ////////////////////////////////////////////////////////////////////

// SPI Transfer That Uses Bitbanging to Send Data
void spitransfer(unsigned char var) {
  digitalWrite(g_sspin, LOW);
  digitalWrite(g_sclk, HIGH);

  for (uint8_t i = 1; i <= 8; i++) {
    digitalWrite(g_sclk, LOW);
    delayMicroseconds(SPIDELAY);
    if (var & 0x80)
      digitalWrite(g_mosi, HIGH);
    else
      digitalWrite(g_mosi, LOW);
    var <<= 1;
    digitalWrite(g_sclk, HIGH);
    delayMicroseconds(SPIDELAY);
  }
  digitalWrite(g_sspin, HIGH);
}

// COMMAND FUNCTIONS ///////////////////////////////////////////////////////////////////////////////

// Sends a Specific Hex Command to the LCD for Execution
void sendcommand(unsigned char command) {
  switch (g_interface) {
    case RS232:
      Serial.write(0xFE);
      Serial.write(command);
      delayMicroseconds(RS232DELAY);
      break;
    case I2C:
      Wire.beginTransmission(g_i2c_address);
      Wire.write(0xFE);
      Wire.write(command);
      Wire.endTransmission();
      delayMicroseconds(I2CDELAY);
      break;
    case SPI:
      spitransfer(0xFE);
      delayMicroseconds(SPIDELAY);
      spitransfer(command);
      delayMicroseconds(SPIDELAY);
      break;
  }
}

// Sends a Specific Command & Associated Value to LCD for Execution
void sendcommand2(unsigned char command, unsigned char value) {
  switch (g_interface) {
    case RS232:
      Serial.write(0xFE);
      Serial.write(command);
      Serial.write(value);
      delayMicroseconds(RS232DELAY);
      break;
    case I2C:
      Wire.beginTransmission(g_i2c_address);
      Wire.write(0xFE);
      Wire.write(command);
      Wire.write(value);
      Wire.endTransmission();
      delayMicroseconds(I2CDELAY);
      break;
    case SPI:
      spitransfer(0xFE);
      delayMicroseconds(SPIDELAY);
      spitransfer(command);
      delayMicroseconds(SPIDELAY);
      spitransfer(value);
      delayMicroseconds(SPIDELAY);
      break;
  }
}

// Allows the User to Send a Direct HD44780 Hex Command to the LCD Display for Execution
void LC_NKC::command(uint8_t value) {
  sendcommand2(COMMAND, value);
}

// RS232 SERIAL FUNCTIONS //////////////////////////////////////////////////////////////////////////

// Displays the Set Baud Rate for the Display
void LC_NKC::displayBaudRate() {
  sendcommand(DISPLAYBAUDRATE);
}

// Changes the RS232 Serial Baud Rate of the LCD
void LC_NKC::changeBaudRate(unsigned char baud) {
  sendcommand2(CHANGEBAUDRATE, baud);
}

// I2C RELATED FUNCTIONS ///////////////////////////////////////////////////////////////////////////

// Displays the Current I2C Address of the LCD
void LC_NKC::displayI2CAddress() {
  sendcommand(DISPLAYI2CADDRESS);
}

// Changes the LCD Display's Default or Current I2C Address
void LC_NKC::changeI2CAddress(unsigned char addr) {
  sendcommand2(CHANGEI2CADDRESS, addr);
}

// Allows the User to Modify the I2C Clock Speed
void changeI2CClock(uint32_t clock_rate) {
  switch (clock_rate) {
    case I2C_SLOW:
      Wire.setClock(I2C_SLOW);
      g_i2c_clock = I2C_SLOW;
      break;
    case I2C_STANDARD:
      Wire.setClock(I2C_STANDARD);
      g_i2c_clock = I2C_STANDARD;
      break;
    case I2C_FAST:
      Wire.setClock(I2C_FAST);
      g_i2c_clock = I2C_FAST;
      break;
  }
}

////////////////////////////////////////////////////////////////////////////////////////////////////
// SOFTWARE LEVEL FUNCTIONS ////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////

// FIRMWARE FUNCTIONS //////////////////////////////////////////////////////////////////////////////

// Displays the Firmware Version of the LCD
void LC_NKC::displayVersion() {
  sendcommand(DISPLAYVERSION);
}

// LCD INITIALIZATION FUNCTIONS ////////////////////////////////////////////////////////////////////

// Turn the NKC LCD Display Pixels & Backlight "ON" or "OFF"
void LC_NKC::on() {
  sendcommand(ON);
  sendcommand2(SETBLBRIGHTNESS, MAX_BRIGHTNESS);
}
void LC_NKC::off() {
  sendcommand(OFF);
  sendcommand2(SETBLBRIGHTNESS, MIN_BRIGHTNESS);
}
// LiquidCrystal Compatible Versions
void LC_NKC::display() {
  on();
}
void LC_NKC::noDisplay() {
  off();
}

// Turn the NKC LCD Display Pixels "ON" or "OFF"
void LC_NKC::displayOn() {
  setContrast(SC_DEF_CONTRAST);
}
void LC_NKC::displayOff() {
  setContrast(MIN_CONTRAST);
}

// LiquidCrystal Compatible Functions
// Initializes the LCD Display and Defaults All Basic Values
void LC_NKC::begin() {
  delay(500);
  switch (g_interface) {
    case RS232:
      Serial.begin(g_baudrate);
      break;
    case I2C:
      Wire.begin();
      break;
    case SPI:
      pinMode(g_sspin, OUTPUT);
      pinMode(g_sclk, OUTPUT);
      pinMode(g_mosi, OUTPUT);
      pinMode(g_miso, INPUT);
      digitalWrite(g_sspin, HIGH);
      break;
  }

  displayOn();
  blinkOff();
  cursorOff();
  clear();
  setContrast(SC_DEF_CONTRAST);
  backlight();
  home();
}
void LC_NKC::init() {
  begin();
}

// BACKLIGHT & CONTRAST FUNCTIONS //////////////////////////////////////////////////////////////////

// LiquidCrystal Compatible Functions
// Turns LCD Backlight "ON" or "OFF"
void LC_NKC::backlight() {
  sendcommand2(SETBLBRIGHTNESS, MAX_BRIGHTNESS);
}
void LC_NKC::noBacklight() {
  sendcommand2(SETBLBRIGHTNESS, MIN_BRIGHTNESS);
}

// Set Backlight Brightness to a Specific Value
void LC_NKC::setBacklight(uint8_t new_val) {
  // To be LCD API compliant, "new_val" must be scaled down
  // from 0-255 to 0-8 by dividing "new_val" by 28.444 (28)
  sendcommand2(SETBLBRIGHTNESS, (new_val / SCALE_BRIGHTNESS));
}
// Alias Function Included For Compatibility
void LC_NKC::setBacklightBrightness(uint8_t new_val) {
  setBacklight(new_val);
}

// Sets the Contrast of the LCD to a Specific Value
void LC_NKC::setContrast(uint8_t new_val) {
  // To be LCD API compliant, "new_val" must be scaled down
  // from 0-255 to 0-50 by dividing "new_val" by 5.019 (5)
  sendcommand2(SETCONTRAST, (new_val / SCALE_CONTRAST));
}

// WRITE, CREATE & PRINT FUNCTIONS /////////////////////////////////////////////////////////////////

// "Writes" Data Over The Specified Interface
size_t LC_NKC::write(uint8_t value) {
  switch (g_interface) {
    case RS232:
      Serial.write(value);
      break;
    case I2C:
      Wire.beginTransmission(g_i2c_address);
      Wire.write(value);
      Wire.endTransmission();
      break;
    case SPI:
      spitransfer(value);
      break;
  }
}


// Creates a New Character & Stores It In LCD Memory
void LC_NKC::createChar(unsigned char char_num, unsigned char *rows) {
  uint8_t i;

  switch (g_interface) {
    case RS232:
      Serial.write(0xFE);
      Serial.write(LOADCUSTOMCHAR);
      Serial.write(char_num);
      Serial.write(rows, CUSTOM_CHAR_SIZE);
      delayMicroseconds(RS232DELAY);
      break;
    case I2C:
      Wire.beginTransmission(g_i2c_address);
      Wire.write(0xFE);
      Wire.write(LOADCUSTOMCHAR);
      Wire.write(char_num);
      Wire.write(rows, CUSTOM_CHAR_SIZE);
      Wire.endTransmission();
      delayMicroseconds(I2CDELAY);
      break;
    case SPI:
      spitransfer(0xFE);
      delayMicroseconds(SPIDELAY);
      spitransfer(LOADCUSTOMCHAR);
      delayMicroseconds(SPIDELAY);
      spitransfer(char_num);
      delayMicroseconds(SPIDELAY);
      for (i = 0; i < CUSTOM_CHAR_SIZE; i++) {
        spitransfer(rows[i]);
        delayMicroseconds(SPIDELAY);
      }
      break;
  }
}

// LiquidCrystal Compatible Functions
// These Functions "print()" Characters to the LCD
void LC_NKC::print(const String inputStr) {

  uint8_t len = inputStr.length();

  switch (g_interface) {
    case RS232:
      for (uint8_t i = 0; i < len; i++) {
        Serial.write(inputStr.charAt(i));
      }
      break;
    case I2C:
      Wire.beginTransmission(g_i2c_address);
      for (uint8_t i = 0; i < len; i++) {
        Wire.write(inputStr.charAt(i));
      }
      Wire.endTransmission();
      break;
    case SPI:
      /*
			for (uint8_t i = 0; i < len; i++)	{
				spitransfer(inputStr.charAt(i));
			}
		  */
      break;
  }
}
void LC_NKC::print(const char inputChar[]) {
  byte len;

  len = min(strlen(inputChar), MAX_STRLEN);
  switch (g_interface) {
    case RS232:
      while (len--)
        Serial.write(*inputChar++);
      break;
    case I2C:
      Wire.beginTransmission(g_i2c_address);
      while (len--)
        Wire.write(*inputChar++);
      Wire.endTransmission();
      break;
    case SPI:
      while (len--)
        spitransfer(*inputChar++);
      break;
  }
}


// These Aliases are Included for Compatibility
void LC_NKC::printstr(const String inputStr) {
  print(inputStr);
}
void LC_NKC::printstr(const char inputChar[]) {
  print(inputChar);
}

// LCD CLEAR FUNCTIONS /////////////////////////////////////////////////////////////////////////////

// LiquidCrystal Compatible Function
// Clears The Entire LCD Display
void LC_NKC::clear() {
  sendcommand(CLEAR);
  delayMicroseconds(CLEARDELAY);
}
// Clears a Single Character by "Column, Row"
void LC_NKC::clear(uint8_t col, uint8_t row) {
  setCursor(col, row);
  print(" ");
  delayMicroseconds(CLEARDELAY);
}
// Clears Specified Columns/Characters of a Row
void LC_NKC::clear(uint8_t start_col, uint8_t end_col, uint8_t row) {
  for (uint8_t i = 0; i <= (end_col - start_col); i++) {
    setCursor((start_col + i), row);
    print(" ");
    delayMicroseconds(CLEARDELAY);
  }
}
// Clears the Contents of the Specified Row
void LC_NKC::clear(uint8_t row) {
  for (uint8_t i = 0; i <= g_num_col; i++) {
    setCursor(i, row);
    print(" ");
    delayMicroseconds(CLEARDELAY);
  }
}

// LCD CURSOR CONTROL FUNCTIONS ////////////////////////////////////////////////////////////////////

// LiquidCrystal Compatible Function
// Quick and Easy Way to Reset Cursor to "0, 0"
void LC_NKC::home() {
  sendcommand(HOME);
}


// LiquidCrystal Compatible Function
// Uses the Hex Index To Set Cursor Position to "Column, Row"
void LC_NKC::setCursor(uint8_t col_num, uint8_t row_num) {
  sendcommand2(SETCURSOR, g_index[row_num] + col_num);
}


//Turn the Underline Cursor "ON" or "OFF"
void LC_NKC::cursorOn() {
  sendcommand(CURSOR_ON);
}
void LC_NKC::cursorOff() {
  sendcommand(CURSOR_OFF);
}
// LiquidCrystal Compatible Versions
void LC_NKC::cursor() {
  sendcommand(CURSOR_ON);
}
void LC_NKC::noCursor() {
  sendcommand(CURSOR_OFF);
}

//Turn the Blinking Cursor "ON" or "OFF"
void LC_NKC::blinkOn() {
  sendcommand(BLINK_ON);
}
void LC_NKC::blinkOff() {
  sendcommand(BLINK_OFF);
}
void LC_NKC::blink(bool isOn) {
  if (isOn == true) {
    sendcommand(BLINK_ON);
  } else if (isOn == false) {
    sendcommand(BLINK_OFF);
  }
}
// LiquidCrystal Compatible Versions
void LC_NKC::blink() {
  sendcommand(BLINK_ON);
}
void LC_NKC::noBlink() {
  sendcommand(BLINK_OFF);
}


//Move the Cursor Right/Left 1 Position (Or More...)
void LC_NKC::left() {
  sendcommand(LEFT);
}
void LC_NKC::right() {
  sendcommand(RIGHT);
}
void LC_NKC::left(uint8_t positions, uint8_t delay_ms) {
  for (uint8_t i = 0; i <= positions; i++) {
    delay(delay_ms);
    sendcommand(LEFT);
  }
}
void LC_NKC::right(uint8_t positions, uint8_t delay_ms) {
  for (uint8_t i = 0; i <= positions; i++) {
    delay(delay_ms);
    sendcommand(RIGHT);
  }
}

//Backspace
void LC_NKC::backspace() {
  sendcommand(BACKSPACE);
}

// OTHER LCD FUNCTIONS /////////////////////////////////////////////////////////////////////////////

//Shift Display Right/Left 1 Position (Or More...)
void LC_NKC::shiftLeft() {
  sendcommand(SHIFTLEFT);
}
void LC_NKC::shiftRight() {
  sendcommand(SHIFTRIGHT);
}
void LC_NKC::shiftLeft(uint8_t positions, uint8_t delay_ms) {
  for (uint8_t i = 0; i <= positions; i++) {
    delay(delay_ms);
    sendcommand(SHIFTLEFT);
  }
}
void LC_NKC::shiftRight(uint8_t positions, uint8_t delay_ms) {
  for (uint8_t i = 0; i <= positions; i++) {
    delay(delay_ms);
    sendcommand(SHIFTRIGHT);
  }
}
void LC_NKC::scrollDisplayLeft(uint8_t positions, uint8_t delay_ms) {
  for (uint8_t i = 0; i <= positions; i++) {
    delay(delay_ms);
    sendcommand(SHIFTLEFT);
  }
}
void LC_NKC::scrollDisplayRight(uint8_t positions, uint8_t delay_ms) {
  for (uint8_t i = 0; i <= positions; i++) {
    delay(delay_ms);
    sendcommand(SHIFTRIGHT);
  }
}

// LiquidCrystal Compatible Versions
void LC_NKC::scrollDisplayLeft() {
  sendcommand(SHIFTLEFT);
}
void LC_NKC::scrollDisplayRight() {
  sendcommand(SHIFTRIGHT);
}